docs/development/design/fsm.md

```diff
@@ -0,0 +1,57 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> IDLE
+    IDLE --> AUTH: "login"
+    AUTH --> READY: "login_success"
+    AUTH --> IDLE: "login_failed"
+    READY --> PEERING: "start"
+    READY --> IDLE: "logout"
+    PEERING --> READY: "stop"
+    PEERING --> READY: "peering_error"
+    PEERING --> IDLE: "authentication_error"
+    PEERING --> IDLE: "logout"
+```
+
+nexd may be in one of four states:
+
+1. IDLE: No configuration has been provided
+2. AUTH: Performing authentication, either interactively or via token
+3. READY: Authenticated and ready to peer
+4. PEERING: Actively Peering
+
+To provide information to the user about the current state, a `nexctl nexd status` command will be added. This can inform the user of the current state and what to do next.
+
+To transition from IDLE to AUTH the "login" transition must occur.
+
+The trigger for this will be a CLI command: `nexctl nexd login`
+
+Once in the "AUTH" state, credentials will be taken from the user - or the device-auth flow will occur. In event of successful login the "login_success" transition occurs to bring nexd into the READY state. In case of "login_failed" we go back to IDLE.
+
+When in the READY state we will immediately issue a transition "start" into the PEERING state to preserve the existing behavior. The commands `nexctl nexd start` and `nexctl nexd stop` will be added to manually issue these transitions so users can choose to start/stop peering as necessary.
```

Comment:
The value of the start/stop commands isn't clear to me. How is it better than `systemctl start` and `systemctl stop` (or the equivalent for whatever system you're on) ?

I saw mention of having to go back through the auth flow, but that's not really the case since we've either got a cached API token from interactive auth or we have user/password. Either way, stopping/starting shouldn't be a big deal, and it's the workflow I would expect based on how most services work.

Even if we did have start/stop, I would expect people actually stop/start the daemon.

Reply:
> The value of the start/stop commands isn't clear to me. How is it better than systemctl start and systemctl stop (or the equivalent for whatever system you're on) ?

Because it allows us a better UX.

On any system with a GUI we're likely going to want a toolbar component and to send notifications to a user.
That toolbar component will want a start/stop button to enable/disable the mesh. fork/exec out to systemctl or the OS'equivalent isn't a particularly elegant or portable way to control our daemon process vs. using our own API.
<img width="347" alt="Screenshot 2023-05-02 at 13 58 52" src="https://user-images.githubusercontent.com/3781777/235673198-927cd488-655e-496a-9986-cee811ba11c7.png">
(see: Tailscale as reference)

Secondly, the control plane connection is used for far more than just getting peers, for example, getting a users current organization memberships, devices etc... We should support cases where a user wants to stop the mesh, browse the list of orgs they are part of, and chose to move their device to another organization

> I saw mention of having to go back through the auth flow, but that's not really the case since we've either got a cached API token from interactive auth or we have user/password. 

Well auth token handling is a topic for another design doc, but point taken. The comment was specifically on recommending a user to use `login/logout` to achieve the use case I'd mentioned - pause VPN while they do something on an overlapping IP address range, or don't want their corporate overlords seeing their packets 😆

Reply:
OK - thanks for adding some more detail. Your example here was incredibly helpful:

> Secondly, the control plane connection is used for far more than just getting peers, for example, getting a users current organization memberships, devices etc... We should support cases where a user wants to stop the mesh, browse the list of orgs they are part of, and chose to move their device to another organization

I think I'm OK with adding this, as long as it's not a hard requirement at startup. In other words, just starting the service should automatically push it to login, assuming it has credentials, or to begin the OTP-login flow. That way, I see the added value while also not making anything harder than it is today.